File: src/main/java/ro/fasttrackit/homework/CountryService.java
package ro.fasttrackit.homework;

import java.util.*;

import static java.util.stream.Collectors.*;


public class CountryService {

    private final List<Country> countries;

    public CountryService(List<Country> countries) {
        this.countries = Optional.ofNullable(countries)
                .map(ArrayList::new)
                .orElseGet(ArrayList::new);
    }

    public List<Country> getCountries() {
        return new ArrayList<>(countries);
    }

    public List<String> countryNames() {
        return countries.stream()
                .map(Country::getName)
                .collect(toList());
    }

    public String countryCapital(String countryName) {
        return countries.stream()
                .map(country -> findCountryByName(countryName).getCapital())
                .findAny()
                .orElse(Optional.empty().toString());
    }

    public Long populationOfCountry(String countryName) {
        return countries.stream()
                .map(country -> findCountryByName(countryName).getPopulation())
                .findAny()
                .orElse(-1L);
    }

    public List<Country> countriesOnContinent(String continent) {
        return countries.stream()
                .filter(country -> isFromContinent(country, continent))
                .collect(toList());
    }

    public List<String> countryNeighbours(String countryName) {
        return countries.stream()
                .map(country -> findCountryByName(countryName).getNeighbours())
                .findAny()
                .orElseGet(ArrayList::new);
    }

    public List<Country> continentCountriesWithMinimumPopulation(String continent, Long minPopulation) {
        return countries.stream()
                .filter(country -> isFromContinent(country, continent))
                .filter(country -> country.getPopulation() >= minPopulation)
                .collect(toList());
    }

    public List<Country> continentCountriesSortedByPopulation(String continent) {
        return countries.stream()
                .filter(country -> isFromContinent(country, continent))
                .sorted(Comparator.comparing(Country::getPopulation))
                .collect(toList());
    }

    public List<Country> continentCountriesSortedByArea(String continent) {
        return countries.stream()
                .filter(country -> isFromContinent(country, continent))
                .sorted(Comparator.comparing(Country::getArea))
                .collect(toList());
    }

    public Map<String, Long> countryAndPopulation(String countryName) {
        return countries.stream()
                .map(country -> findCountryByName(countryName))
                .collect(toMap(Country::getName, Country::getPopulation, (v1, v2) -> v1));
    }

    public Map<String, Long> countriesToPopulationsMap() {
        return countries.stream()
                .collect(toMap(Country::getName, Country::getPopulation));
    }

    public Map<String, List<Country>> countriesToContinentsMap() {
        return countries.stream()
                .collect(groupingBy(Country::getContinent));
    }

    public Map<String, List<Country>> countriesToContinentsSortedByPopulationMap() {
        return countries.stream()
                .sorted(Comparator.comparing(Country::getPopulation))
                .collect(groupingBy(Country::getContinent));
    }

    public Map<String, List<Country>> countriesFromContinent(String continent) {
        return countries.stream()
                .filter(country -> isFromContinent(country, continent))
                .collect(groupingBy(Country::getContinent));
    }

    public Map<String, List<Country>> countriesFromContinentSortedByPopulation(String continent) {
        return countries.stream()
                .filter(country -> isFromContinent(country, continent))
                .sorted(Comparator.comparing(Country::getPopulation))
                .collect(groupingBy(Country::getContinent));
    }

    public List<Country> countriesByNeighbours(String neighbouring, String avoiding) {
        return countries.stream()
                .filter(country -> country.getNeighbours().contains(neighbouring))
                .filter(country -> !country.getNeighbours().contains(avoiding))
                .collect(toList());
    }

    private boolean isFromContinent(Country country, String continent) {
        return country.getContinent().equalsIgnoreCase(continent);
    }

    private Country findCountryByName(String countryName) {
        return countries.stream()
                .filter(country -> country.getName().equalsIgnoreCase(countryName))
                .findAny()
                .orElse(new Country("default", "default", 0L, 0L, "default", List.of()));
    }

}

File: src/main/java/ro/fasttrackit/kotlinhomework/CountryService.kt
package ro.fasttrackit.kotlinhomework

data class Country(val name: String, val capital: String, val population: Long,
                   val area: Long, val continent: String, val neighbours: List<String>) {
    val validator = CountryValidator().verify(name, capital, population, area, continent)
}

class CountryService(private val countries: List<Country>) {

    fun getAllCountries(): List<Country> = countries

    fun countryNames(): List<String> = countries.map { it.name }

    fun countryCapital(countryName: String): String? = countries
            .find { it == getCountryByName(countryName) }?.capital

    fun countryPopulation(countryName: String): Long? = countries
            .find { it == getCountryByName(countryName) }?.population

    fun continentCountries(continent: String) = countries.filter { isFromContinent(it, continent) }

    fun countryNeighbours(countryName: String): List<String>? = countries
            .find { it == getCountryByName(countryName) }?.neighbours

    fun continentCountriesWithMinimumPopulation(continent: String, minimumPopulation: Long) = countries
            .filter { isFromContinent(it, continent) }
            .filter { it.population >= minimumPopulation }

    fun continentCountriesSortedByPopulation(continent: String) = countries
            .filter { isFromContinent(it, continent) }
            .sortedBy { it.population }

    fun continentCountriesSortedByArea(continent: String) = countries
            .filter { isFromContinent(it, continent) }
            .sortedBy { it.area }

    fun countryAndPopulationMap(countryName: String): Map<String, Long> = countries
            .filter { isCountry(it, countryName) }
            .associate { it.name to it.population }

    fun countriesAndPopulationMap(): Map<String, Long> = countries.associate { it.name to it.population }

    fun countriesFromContinentMap(continent: String): Map<String, List<Country>> = countries
            .filter { isFromContinent(it, continent) }
            .groupBy { it.continent }

    fun countriesToContinentsMap(): Map<String, List<Country>> = countries.groupBy { it.continent }

    fun countriesToContinentSortedByPopulationMap(): Map<String, List<Country>> = countries
            .sortedBy { it.population }
            .groupBy { it.continent }

    fun countriesFromContinentSortedByPopulationMap(continent: String): Map<String, List<Country>> = countries
            .filter { isFromContinent(it, continent) }
            .sortedBy { it.population }
            .groupBy { it.continent }

    fun countriesByNeighbours(neighbouring: String, avoiding: String): List<Country> = countries
            .filter { it.neighbours.contains(neighbouring) }
            .filter { !it.neighbours.contains(avoiding) }

    private fun isFromContinent(country: Country, continent: String) = country.continent.equals(continent, ignoreCase = true)

    private fun isCountry(country: Country, countryName: String) = country.name.equals(countryName, ignoreCase = true)

    private fun getCountryByName(countryName: String): Country = countries.first { it.name.equals(countryName, ignoreCase = true) }
}